---
title: "Microbial Mapping - Summary"
output: html_notebook
---

# Document Summary
This document is a guide to the different metadata categories, sample numbers, sampling sites, etc. that are included in the microbial mapping dataset. 

The goal is to give a good introudction to the data collection and project structure.
The final figure is a summary of all the samples that are included in the sequencing data. 

Code does not require abundance files and can be run locally.

## Libraries etc.
Load some basic data manipulation and visualization libraries.
Set a standardized custom color scale for visualizations.
```{r}
library(dplyr)
library(ggplot2)
library(data.table)
library(ggpubr)


# set ggplot theme
theme_set(theme_bw())

# set color palettes 
pal_d = "Spectral"
scale_colour_discrete <-  function(palname=pal_d, ...){
  scale_colour_brewer(palette=palname, ...)
}
scale_fill_discrete <-  function(palname=pal_d, ...){
    scale_fill_brewer(palette=palname, ...)
}

pal_c = "C"
uniform_gradient_color <- function(palname = pal_c){
  scale_color_viridis_c(option = palname)
}

source("../src/get_colors.R")

trophic_colors <- get_waimea_colors(cat_type = "trophic")
habitat_colors <- get_waimea_colors(cat_type = "habitat")
emp_colors     <- get_waimea_colors(cat_type = "emp") 

```

## Reading in the metadata
The metadata has been split into unique sample metadata and preparation information. This is
in accordance with the EMP data uploading guidelines. 
The 'sample_name' column links sample prep and collection metadata.

### Preparation Information

- *sample_name* = Unique integer coresponding to a sample taken from the field. Exception: controls duplicated across sequencing runs were given unique IDs for the same sample. Matches sample metadata column.
- *barcode* = 12-mer index from EMP aliquots. Unique per sample per run.
- *linker* = Primer linker sequences.
- *primer* = Primer sequence of barcoded primer.
- *extraction_kit* = DNA extraction kit used.
- *run_prefix* = Filename prefix for raw sequencing data.
- *locus* = Locus for amplification.
- *target_gene* = Target gene name.
- *target_subfragment* = Subfragment of gene targeted. 
- *sequencing_id* = Unique integer ID for each sequenced well. 
- *primer_plate* = Integer indicating primer plate that was used for amplification.
- *primer_well* = Well id of primer plate. 
- *forward_primer* = Forward primer sequence.
- *reverse_primer* = Reverse primer sequence.
- *primer_name* = Primer name assigned to well. 
- *pcr_plate* = Name of PCR plate from which sequence product originated.
- *platform* = Sequencing platform. 
- *instrument_model* = Sequencing instrument model. 


### Sample Data Information

Sample Identifiers:

- *sample_name* = Unique integer coresponding to a sample taken from the field. Exception: controls duplicated across sequencing runs were given unique IDs for the same sample.Matches preparation metadata column.
- *sample_id* = Unique integer combined with some abbreviated metadata information to make it more human readable. Useful for labelling tubes but not necessary for analysis. 
- *collection_label* = Label as it was written on the original sampling tube. Once sample entered the lab, it was accessioned into the spreadsheet and assigned a unique sample_name.
- *project_name* = For microbial mapping, the project should be "Keck". A few samples from other projects were included in this dataset. 
- *geo_loc_name* = Sampling location following earth microbiome project.

Site Identifiers:

- *site_code* = Unique integer code for each sampling site. Paired stream and terrestrial sites have the same code.
- *site_name* = Unique human readable name for each sampling site.
- *site_type*
- *latitude_deg*
- *longitude_deg*

Sample Type: 


- *sample_type* = Categorical variable unique to this project. Sampling at each site targeted these 'sample types', which were sampled as consistently as possible across sites. 
- *habitat* = Categorical variable unique to this project indicating one of three habitat types.  
- *host* = Categorical variable unique to this project indicating whether sample was host associated and broad host types. 


Individual Sample:
- *trophic* = Categorical variable unique to this project indicating trophic status of each sample. Determined individually for each sample and sometimes varies within a sample type. 
- *metadata* = Any other metadata for this sample. Equivalent to the "Description" column for qiime. 
- *common_name* = Common name for collected sample used for human readable summaries. 
- *family* = Family level taxonomy for host associated samples, when available.
- *genus* = Genus level taxonomy for host associated samples, when available.
- *species* = Species level taxonomy for host associated samples, when available.
- *taxonomy_info* = State of taxnomic id.
- *collection_timestamp* = Date of sample collection


EMP metadata
- *empo_1* = Categorical variable following earth microbiome project.
- *empo_2* = Categorical variable following earth microbiome project.
- *empo_3* = Categorical variable following earth microbiome project.
- *env_material* = Categorical variable following earth microbiome project.
- *env_feature* = Categorical variable following earth microbiome project.
- *env_biome* = Categorical variable following earth microbiome project.
- *enve_biom_0* = Categorical variable following earth microbiome project.
- *envo_biom_1* = Categorical variable following earth microbiome project.
- *envo_biome_2* = Categorical variable following earth microbiome project.
- *envo_biome_3* = Categorical variable following earth microbiome project.

Sample Processing
- *accessioned* = Sample management variable used to check whether the sample was accessioned once the entry had been created.
- *PI* = Designated project PI responsible for sampling protocol.
- *Processing* = Sample management variable used to keep track of additional sample processing prior to extraction. 

## Data Clean Up
Some samples in this data set are not associated with the Waimea project.
Filter those samples out.

```{r}
# read in
samples <- fread("../data/raw/sample/unique_waimea_hiseq_samples_emp.tsv")

# projects
unique(samples$project_name)

# filter by project
samples <- samples[project_name %in% c("Keck","JoshMS") & site_name != "Lab",]

# count all beach samples as riverine
samples[site_name == "Beach" & sample_type == "WaterSaline", habitat:= "Riverine"]

# simplify trophic relationships
samples[!(trophic %in% c("Environmental", "PrimaryProducer")), trophic := "Consumer"]

```


## Broad Summary
Certain sample types were collected for each site in a given habitat.
Summarize the sample types in each habitat and by each site to make sure everything looks clean.


```{r}
# sample counts of each type by habitat
samples[ ,
         .N,
         by= .(habitat)]

# sample counts by site
site_sum <- samples[ ,
                       .N,
                       by = .(site_name)]


capture.output(print(paste0("Global Sample Count Mean = ", mean(site_sum$N), "SD = ", sd(site_sum$N))),
               file = "../data/processed/global_mean_sample_per_plot.txt")

# sample count summary by habitat and site
hab_site_sum <- samples[ ,
         .N,
         by = .(habitat, site_name)]



hab_site_means <- hab_site_sum[ ,
                               .(mean_sample_count = mean(N), sd = sd(N)),
                               by = habitat]

fwrite(x = hab_site_means, "../data/processed/hab_site_means.csv")

```

## Trophic Summary
One of the main goals of this project is to track microbial communit characteristics across trophic levels.

```{r}
# summarise sample counts by trophicl level and habitat
troph_by_habitat <- samples[ , .N, keyby = .(habitat, trophic)]

# barplot of counts by habitat, stacked trophic levels
b1 <- ggplot(troph_by_habitat, aes(y = N, x = habitat, fill = trophic )) +
  geom_bar(position="stack", stat="identity")+
  labs(x = "Habitat", y = "Total # Samples")+
  scale_fill_manual(values = trophic_colors)

b1
ggsave("total_samples_per_habitat.pdf")

```


## Barplots


```{r}
## normal barplots
# count N observations for columns
troph_data <- samples[ , .N, by = .(habitat, common_name, trophic)]

# assign a new placeholder columns group and key
troph_data$group <- troph_data$habitat
troph_data$key <- paste(troph_data$habitat, troph_data$common_name)

# filter out samples with no habitat (controls)
troph_data <- na.omit(troph_data, cols="habitat")

## generate unique ids for each bar
ids <- troph_data[ , .(tot=sum(N)), .(key,habitat)][order(habitat,tot)]

ids$id <- 1:nrow(ids)

# join ids with the rest of trophic data
setkey(troph_data, habitat, key)
setkey(ids, habitat,key) 
troph_data <- merge(troph_data,ids, all.x = T)


troph_data <- troph_data[order(habitat, tot)]

habitats <- unique(troph_data$habitat)


# make a summary barplot for each habitat
b_plots <- list()

for (a in seq_along(habitats)) {
  
  habidata <- troph_data[troph_data$habitat == habitats[a],]
  
b_plots[[a]] <- ggplot(data=habidata, aes(x=reorder(common_name,id), y = N, fill=trophic)) +
  geom_bar(stat = "identity") +
  labs(fill ="Trophic Level")+
  scale_fill_manual(values = trophic_colors)+
  coord_flip()+
  labs(y = "", x = "", title = habitats[a])+
  ylim(0,150)

print(b_plots[[a]])
ggsave(file = paste("outputs/summary/habitat_", habitats[a],"_barplot.pdf",sep = ""),
       plot = b_plots[[a]], device = "pdf", width = 10, height = 10, dpi = 400)

}

# make faceted plot

# convert plot to grob
bar_grobs <-lapply(b_plots, ggplotGrob)

# set a standard width for all grobs
std_width <- bar_grobs[[1]]$widths
bar_std <- lapply(bar_grobs, function(x) {
                  x$widths <- std_width
                  return(x)})
# arrange ordinations
bar_g <- ggarrange(plotlist=bar_std,
                nrow = 1, ncol = 3)


ggsave(paste0("outputs/summary/all_samples_facet_barplot.pdf"),
       plot = bar_g, 
       width = 15,
       height = 5)


```


## Circle Barplot
See RGraph Gallery for more detailed explanation on how this works.

```{r}
# calculate the ANGLE of the labels
label_data <- troph_data[ , .(tot=sum(N)), by= .(id,common_name)]

number_of_bar <- nrow(label_data)
angle <- 90 - 360 * (label_data$id-0.5) /number_of_bar     # I substract 0.5 because the letter must have the angle of the center of the bars. Not extreme right(1) or extreme left (0)
label_data$hjust <- ifelse( angle < -90, 1, 0)
label_data$angle <- ifelse(angle < -90, angle+180, angle)

# prepare a data frame for base lines
base_data <- troph_data %>% 
  group_by(group) %>% 
  summarize(start=min(id), end=max(id)) %>% 
  rowwise() %>% 
  mutate(title=mean(c(start, end)))

# prepare a data frame for grid (scales)
grid_data <- base_data
grid_data$end <- grid_data$end[ c( nrow(grid_data), 1:nrow(grid_data)-1)] + 1
grid_data$start <- grid_data$start - 1
grid_data <- grid_data[-1,]


# Limits of the plot = very important. The negative value controls the size of the inner circle, the positive one is useful to add size over each bar
p <-
  ggplot(data = troph_data, aes(x = factor(id), y = N, fill = trophic)) +
  geom_bar(stat = "identity") +
  labs(fill = "Trophic Level") +
  scale_fill_manual(values = trophic_colors) +
  
  ylim(-100, 210) +
    
  # Custom the theme: no axis title and no cartesian grid
  theme_minimal(base_size = 20) +
  
  # remove some plot elements
  theme(
    axis.text = element_blank(),
    axis.title = element_blank(),
    panel.grid = element_blank(),
    plot.margin = unit(rep(-1, 4), "cm")# Adjust the margin to make in sort labels are not truncated!
  ) +
#   # 
#   # # This makes the coordinate polar instead of cartesian.
   coord_polar(start = 0)  +
  # add base line segments for each habitat
  geom_segment(
    data = base_data,
    aes(
      x = start,
      y = -5,
      xend = end,
      yend = -5
    ),
    colour = "black",
    alpha = 0.8,
    size = 0.6 ,
    inherit.aes = FALSE
  )  +
  # add habitat text labels to baselines
  geom_text(
    data = base_data,
    aes(x = title - 5, y = -22, label = group),
    hjust = c(.8, .8, .2),
    colour = "black",
    alpha = 0.8,
    size = 6,
    fontface = "bold",
    inherit.aes = FALSE
    ) +

#   # # Add the labels, using the label_data dataframe that we have created before
geom_text(data=label_data,
          aes(x=factor(id), y=tot+10, label=common_name, hjust=hjust),
          color="black", fontface="bold",alpha=0.6, size=6,
          angle= label_data$angle, inherit.aes = FALSE )   
#
p
# 
ggsave("outputs/summary/habitat_sample_type_circleplot.pdf",plot = p, device = "pdf", width = 15, height = 10, dpi = 400)


```


## Sunburst

```{r}
library(plotly)
source("../src/plot_functions.R")

sunburstDF <- as.sunburstDF(troph_data[c(1:4)], valueCol = "value")

fig <- plot_ly(data = sunburstDF, ids = ~ids, labels= ~labels, parents = ~parents, values= ~values, type='sunburst', branchvalues = 'total')

fig

# use orca command line interface to produce static pdf (produces errors, but works anyways)
orca(fig, file = "outputs/summary/summary_sunburst.pdf",format ="pdf") 
```

